### PR TITLE
[WIP] Add StateRefreshFunc when creating apps in spaces

### PR DESCRIPTION
**Pull request by talbright**
Use StateRefreshFunc to check the state of the app (only applicable to apps in spaces.) The API endpoints for `/app` don't include json responses with a `status` field. However, we can look for a specific(not publicly documented) Heroku error code `app_not_provisioned`.

Addresses #71

### CODE REVIEW

File: heroku/resource_heroku_app.go [hunk 1]
```diff
@@ -704,6 +715,29 @@ func performAppPostCreateTasks(d *schema.ResourceData, client *heroku.Service) e
...
+func appStateRefreshFunc(client *heroku.Service, appID string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		appInfo, err := client.AppInfo(context.TODO(), appID)
+		if err != nil {
+			if herr, ok := err.(*url.Error).Err.(heroku.Error); ok {
+				if herr.ID == "app_not_provisioned" {
+					log.Printf("[DEBUG] Still allocating: %s", appID)
+					return nil, "allocating", nil
+				} else {
+					log.Printf("[WARN] Unexpected response waiting for allocation: %s", herr.ID)
```
[the remark this answers is not in the document]
Comment (talbright) [reply heroku/resource_heroku_app.go]: True. I'll remove it.

**Comment by talbright** [comment]
@bernerdschaefer after some testing it looks like the theory regarding the api returning `app_not_provisioned` via `GET /apps/shinjuku-gyoen` isn't holding up. However, it is being returned via `POST /apps/shinjuku-gyoen/domains`:

```
2018-08-16T19:30:42.321-0400 [DEBUG] plugin.terraform-provider-heroku: Host: api.heroku.com
2018-08-16T19:30:42.321-0400 [DEBUG] plugin.terraform-provider-heroku: User-Agent: heroku/v3 (darwin; amd64)
2018-08-16T19:30:42.321-0400 [DEBUG] plugin.terraform-provider-heroku: Content-Length: 48
2018-08-16T19:30:42.321-0400 [DEBUG] plugin.terraform-provider-heroku: Accept: application/vnd.heroku+json; version=3
2018-08-16T19:30:42.321-0400 [DEBUG] plugin.terraform-provider-heroku: Authorization: Basic xxx
2018-08-16T19:30:42.321-0400 [DEBUG] plugin.terraform-provider-heroku: Content-Type: application/json
2018-08-16T19:30:42.321-0400 [DEBUG] plugin.terraform-provider-heroku: Request-Id: 163e3b66-b334-4aff-8b72-4d7a60792172
2018-08-16T19:30:42.321-0400 [DEBUG] plugin.terraform-provider-heroku: Accept-Encoding: gzip
2018-08-16T19:30:42.321-0400 [DEBUG] plugin.terraform-provider-heroku:
2018-08-16T19:30:43.964-0400 [DEBUG] plugin.terraform-provider-heroku: 2018/08/16 19:30:43 response:
2018-08-16T19:30:43.964-0400 [DEBUG] plugin.terraform-provider-heroku: HTTP/1.1 422 Unprocessable Entity
2018-08-16T19:30:43.964-0400 [DEBUG] plugin.terraform-provider-heroku: Content-Length: 124
2018-08-16T19:30:43.964-0400 [DEBUG] plugin.terraform-provider-heroku: Content-Type: application/json; charset=utf-8
2018-08-16T19:30:43.964-0400 [DEBUG] plugin.terraform-provider-heroku: Date: Thu, 16 Aug 2018 23:30:43 GMT
2018-08-16T19:30:43.964-0400 [DEBUG] plugin.terraform-provider-heroku: Request-Id: 51846c8e-f01c-4ba9-b9f4-cfa3af50eab8,163e3b66-b334-4aff-8b72-4d7a60792172,179807b7-0eba-4998-7dfc-e97dd7bae036,c14b9536-3b3d-3f2a-42de-cfed29c8ba03
2018-08-16T19:30:43.964-0400 [DEBUG] plugin.terraform-provider-heroku: Via: 1.1 spaces-router (1645f63f1e8f), 1.1 spaces-router (1645f63f1e8f)
2018-08-16T19:30:43.964-0400 [DEBUG] plugin.terraform-provider-heroku:
2018-08-16T19:30:43.964-0400 [DEBUG] plugin.terraform-provider-heroku:
2018-08-16T19:30:43.964-0400 [DEBUG] plugin.terraform-provider-heroku: 2018/08/16 19:30:43 response body (len=124):
2018-08-16T19:30:43.964-0400 [DEBUG] plugin.terraform-provider-heroku: '{"id":"space_app_not_provisioned","message":"App shinjuku-gyoen is being provisioned. Please try again later.","status":422}'

Maybe we can open a bug report on the API, we would want a `status` field ideally.

Given that, unless you have an alternative suggestion, I'm going to move this refresh func to the `heroku_domain` resource.

**Comment by talbright** [comment]
@davidji99 I talked with @bernerdschaefer about this. A common workaround in the wild for this issue:

```hcl
  provisioner "local-exec" {
    command = "echo 'Waiting for app to be ready..'; sleep 15"
  }
```

What do you think about us adding a provider config option (piggy backing off of #113 when its complete), and then do a `time.Sleep(config.PostAppCreationDelay * time.Second)` in the `heroku_app` resource? Not ideal but until the API is fixed, there's not much we can do....

**Comment by davidji99** [comment]
```
What do you think about us adding a provider config option (piggy backing off of #113 when its complete), and then do a time.Sleep(config.PostAppCreationDelay * time.Second) in the heroku_app resource? Not ideal but until the API is fixed, there's not much we can do....
```

@talbright, So you're thinking something along the lines of an attribute on the provider like the below:
```
provider "heroku" {
  email   = "ops@company.com"
  api_key = "${var.heroku_api_key}"
  wait_for_app_creation = 15
}
```

that then gets eventually passed down to `resource_heroku_app.go`? If so, that would work and might be cleaner than the `local-exec`

**Comment by talbright** [comment]
Exactly.

**Comment by davidji99** [comment]
Gotcha, that sounds pretty flexible to me. Let me finish up my PR.

**Comment by talbright** [comment]
I found this: https://www.terraform.io/docs/configuration/resources.html#timeouts and at first thought we might be able to use that instead, but its not a good fit. So I'll implement this as discussed above now that #113 is in.

**Comment by davidji99** [comment]
Test

**Comment by talbright** [comment]
> Is the reason for not being a good fit because their timeout essentially is a (say) ruby sleep whereas we want to poll every x seconds to see if the app is created?
The way I understand this feature is that it affects timeouts for creation, deletion, etc whereas we want to simply pause at the end of creation.

**Comment by talbright** [comment]
closing this in favor of #142

**Pull request closed without merging**